jina/executors/indexers/keyvalue.py

```diff
@@ -69,12 +70,58 @@ def add(self, keys: Iterator[int], values: Iterator[bytes], *args, **kwargs):
             # print(f'l: {l} p: {p} r: {r} r+l: {r + l} size: {self._size}')
 
     def query(self, key: int) -> Optional[bytes]:
+        # FIXME this shouldn't be required
+        if self.query_handler is None:
+            self.query_handler = self.get_query_handler()
         pos_info = self.query_handler.header.get(key, None)
         if pos_info is not None:
             p, r, l = pos_info
             with mmap.mmap(self.query_handler.body, offset=p, length=l) as m:
                 return m[r:]
 
+    def update(self, keys: Iterator[int], values: Iterator[bytes], *args, **kwargs):
+        new_values = []
+        new_keys = []
+        for key, new_value in zip(keys, values):
+            pos_info = self.query_handler.header.get(key, None)
+            if pos_info is not None:
+                p, r, l = pos_info
+                with mmap.mmap(self.query_handler.body, offset=p, length=l) as m:
+                    old_value = m[r:]
+                    if old_value != new_value:
+                        new_keys.append(key)
+                        new_values.append(new_value)
+
+        if new_keys:
+            adder = self.get_add_handler()
+            _start = adder.body.tell()
+            for key, new_value in zip(new_keys, new_values):
+                l = len(new_value)  #: the length
+                p = int(_start / self._page_size) * self._page_size  #: offset of the page
+                r = _start % self._page_size  #: the remainder, i.e. the start position given the offset
+                adder.header.write(
+                    np.array(
+                        (key, p, r, r + l),
+                        dtype=np.int64
+                    ).tobytes()
+                )
+                _start += l
+                adder.body.write(new_value)
```

Comment:
isn't this self.add()?

Reply:
No, because we need the `adder` handler, and we need a custom `_start` position